src/synchronization_c.c4

```diff
@@ -164,3 +306,103 @@ SHMEM_BIND_C_SYNC(`SHMEM_DEF_WAIT_UNTIL')
+        int ncompleted = 0;                                                                    \
+        SHMEM_ERR_CHECK_INITIALIZED();                                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(vars, sizeof(TYPE));                                         \
+        SHMEM_ERR_CHECK_CMP_OP(cond);                                                          \
+                                                                                               \
+        if (nelems == 0) return 1;                                                             \
+                                                                                               \
+        for (int i = 0; i < nelems; i++) {                                                     \
+            int cmpret;                                                                        \
+            COMP(cond, vars[i], value, cmpret);                                                \
+            if (cmpret) ncompleted++;                                                          \
+        }                                                                                      \
+        if (ncompleted == nelems) {                                                            \
+            shmem_internal_membar_load();                                                      \
+            return 1;                                                                          \
+        } else {                                                                               \
+            return 0;                                                                          \
+        }                                                                                      \
+    }
+
+SHMEM_BIND_C_SYNC(`SHMEM_DEF_TEST_ALL')
+
+
+#define SHMEM_DEF_TEST_ANY(STYPE,TYPE)                                                         \
+    size_t SHMEM_FUNCTION_ATTRIBUTES                                                           \
+    shmemx_##STYPE##_test_any(TYPE *vars, size_t nelems, int * restrict status,                \
+                              int cond, TYPE value)                                            \
+    {                                                                                          \
+        SHMEM_ERR_CHECK_INITIALIZED();                                                         \
+        SHMEM_ERR_CHECK_SYMMETRIC(vars, sizeof(TYPE));                                         \
+        SHMEM_ERR_CHECK_CMP_OP(cond);                                                          \
+                                                                                               \
+        size_t found_idx = SIZE_MAX, i = 0;                                                    \
+        SHMEM_MUTEX_LOCK(shmem_internal_mutex_rand_r);                                         \
+        size_t start_idx = rand_r(&shmem_internal_rand_seed) / (RAND_MAX + 1.0) * nelems;      \
```

Comment:
Does this need to be cast to `size_t`?  If not, then we might not be getting the floating point arithmetic we want here.  If so, this should generate a type conversion warning.

Reply:
We have a couple of implicit casts here:
```
main.c:20:57: warning: conversion to ‘double’ from ‘size_t {aka long unsigned int}’ may alter its value [-Wconversion]
     size_t start_idx = rand_r(&seed) / (RAND_MAX + 1.0) * nelems;
                                                         ^
main.c:20:24: warning: conversion to ‘size_t {aka long unsigned int}’ from ‘double’ may alter its value [-Wfloat-conversion]
     size_t start_idx = rand_r(&seed) / (RAND_MAX + 1.0) * nelems;
```
So it looks like that cast to `size_t` is happening (at least with this older gcc compiler).  But is your concern that we're not getting a perfectly uniform distribution?  A quick test suggests it's pretty darn good:
![rand_r_with_casts](https://user-images.githubusercontent.com/1834947/49468572-430ebe80-f7d3-11e8-9793-3131875c945c.png)

Reply:
The concern is with implicit conversions that can cause loss of precision.  This type of thing can generate warnings with some compilers and probably will pop up in future code scans.  Generally better not to rely this.  I'd update this to:
```
size_t start_idx = (size_t) (rand_r(&seed) / (RAND_MAX + 1.0) * (double) nelems);
```

Reply:
Also, nice histogram.  🏆

Reply:
Excellent!  I even like how the explicit casting looks for some reason.

Yeah, I just _had_ to try github's drag-n-drop image uploading feature.  I only wish I could change the image size easily after uploading.  I didn't intend to make it so large...